sql/sem/eval: TestCastMap failed

sql/sem/eval.TestCastMap [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20492432?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20492432?buildTab=artifacts#/) on master @ [3c9f2eec87cadbd0c2a0fab0692ca552f1553530](https://github.com/cockroachdb/cockroach/commits/3c9f2eec87cadbd0c2a0fab0692ca552f1553530):


```
test logs left over in: /artifacts/tmp/_tmp/e1391e4816aa0a405ecff2c30bb36266/logTestCastMap2052808372
    ftoa.go:68: panic: strconv: illegal AppendFloat/FormatFloat bitSize
--- FAIL: TestCastMap (0.12s)
panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize

goroutine 28 [running]:
testing.tRunner.func1.2({0x2698100, 0x3251240})
	GOROOT/src/testing/testing.go:1632 +0x1bc
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x334
panic({0x2698100?, 0x3251240?})
	GOROOT/src/runtime/panic.go:791 +0x124
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:136 +0x384
panic({0x2698100?, 0x3251240?})
	GOROOT/src/runtime/panic.go:791 +0x124
strconv.genericFtoa({0x0?, 0xed1954?, 0x4000cacb98?}, 0xed1dec?, 0x98?, 0xed19e0?, 0x4000cacb98?)
	GOROOT/src/strconv/ftoa.go:68 +0x524
strconv.AppendFloat(...)
	GOROOT/src/strconv/ftoa.go:54
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.PgwireFormatFloat({0x0?, 0x0, 0xa?}, 0x4000cacd00?, {0xcacbf8?, 0x40?, 0x600ea0?, {0x0?, 0xcacc18?}}, 0x416904?)
	pkg/sql/sem/tree/pgwire_encode.go:215 +0x13c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).pgwireFormat(0x4001001b60, 0x400102a000)
	pkg/sql/sem/tree/pgwire_encode.go:90 +0x19c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).Format(0x4000cace08?, 0xebf160?)
	pkg/sql/sem/tree/datum.go:4512 +0xb8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrAdjustConstants(0x400102a000, {0x325d780, 0x4001001b60})
	pkg/sql/sem/tree/adjust_constants.go:88 +0x384
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x400102a000, {0x325d780, 0x4001001b60})
	pkg/sql/sem/tree/format.go:703 +0x98
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.AsStringWithFlags({0x325d780, 0x4001001b60}, 0x10002, {0x4000cad518?, 0x10?, 0x10?})
	pkg/sql/sem/tree/format.go:841 +0x64
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCastWithoutPrecisionTruncation({0x3287968, 0x4e5c500}, 0x40003f6008, {0x32b5860?, 0x4001001b60?}, 0x4b05080, 0x1)
	pkg/sql/sem/eval/cast.go:453 +0x66e4
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCast({0x3287968?, 0x4e5c500?}, 0x4000015f98?, {0x32b5860?, 0x4001001b60?}, 0x4b05080, 0x38?)
	pkg/sql/sem/eval/cast.go:100 +0x30
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.PerformCast(...)
	pkg/sql/sem/eval/cast.go:65
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap.func1(0x278cae0?, 0x15f98, 0x80?, 0xa4?, 0x36?)
	pkg/sql/sem/eval/cast_map_test.go:54 +0x120
github.com/cockroachdb/cockroach/pkg/sql/sem/cast.ForEachCast(0x4000cadf20)
	pkg/sql/sem/cast/cast.go:158 +0x2dc
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap(0x4000d0e1a0)
	pkg/sql/sem/eval/cast_map_test.go:39 +0x17c
testing.tRunner(0x4000d0e1a0, 0x2e70230)
	GOROOT/src/testing/testing.go:1690 +0xe4
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1743 +0x314
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCastMap.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/sem/eval.TestCastMap [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20492810?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20492810?buildTab=artifacts#/) on master @ [5feef230f17a153f81f2190ffc952c63cc0879d8](https://github.com/cockroachdb/cockroach/commits/5feef230f17a153f81f2190ffc952c63cc0879d8):


```
test logs left over in: /artifacts/tmp/_tmp/e1391e4816aa0a405ecff2c30bb36266/logTestCastMap1987818227
    ftoa.go:68: panic: strconv: illegal AppendFloat/FormatFloat bitSize
--- FAIL: TestCastMap (0.13s)
panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize

goroutine 82 [running]:
testing.tRunner.func1.2({0x2698100, 0x3251240})
	GOROOT/src/testing/testing.go:1632 +0x1bc
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x334
panic({0x2698100?, 0x3251240?})
	GOROOT/src/runtime/panic.go:791 +0x124
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:136 +0x384
panic({0x2698100?, 0x3251240?})
	GOROOT/src/runtime/panic.go:791 +0x124
strconv.genericFtoa({0x0?, 0xed1954?, 0x4000524b98?}, 0xed1dec?, 0x98?, 0xed19e0?, 0x4000524b98?)
	GOROOT/src/strconv/ftoa.go:68 +0x524
strconv.AppendFloat(...)
	GOROOT/src/strconv/ftoa.go:54
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.PgwireFormatFloat({0x0?, 0x0, 0xb?}, 0x4000524d00?, {0x524bf8?, 0x40?, 0x600ea0?, {0x0?, 0x27905e0?}}, 0x4000b6c5d0?)
	pkg/sql/sem/tree/pgwire_encode.go:215 +0x13c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).pgwireFormat(0x4000a9db90, 0x40009a2180)
	pkg/sql/sem/tree/pgwire_encode.go:90 +0x19c
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).Format(0x4000524e08?, 0xebf160?)
	pkg/sql/sem/tree/datum.go:4512 +0xb8
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrAdjustConstants(0x40009a2180, {0x325d780, 0x4000a9db90})
	pkg/sql/sem/tree/adjust_constants.go:88 +0x384
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0x40009a2180, {0x325d780, 0x4000a9db90})
	pkg/sql/sem/tree/format.go:703 +0x98
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.AsStringWithFlags({0x325d780, 0x4000a9db90}, 0x10002, {0x4000525518?, 0x10?, 0x10?})
	pkg/sql/sem/tree/format.go:841 +0x64
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCastWithoutPrecisionTruncation({0x3287968, 0x4e5c500}, 0x4000867808, {0x32b5860?, 0x4000a9db90?}, 0x4b05080, 0x1)
	pkg/sql/sem/eval/cast.go:453 +0x66e4
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCast({0x3287968?, 0x4e5c500?}, 0x4000015f98?, {0x32b5860?, 0x4000a9db90?}, 0x4b05080, 0x38?)
	pkg/sql/sem/eval/cast.go:100 +0x30
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.PerformCast(...)
	pkg/sql/sem/eval/cast.go:65
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap.func1(0x278cae0?, 0x15f98, 0x0?, 0xa6?, 0x31?)
	pkg/sql/sem/eval/cast_map_test.go:54 +0x120
github.com/cockroachdb/cockroach/pkg/sql/sem/cast.ForEachCast(0x4000525f20)
	pkg/sql/sem/cast/cast.go:158 +0x2dc
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap(0x40001cd6c0)
	pkg/sql/sem/eval/cast_map_test.go:39 +0x17c
testing.tRunner(0x40001cd6c0, 0x2e70230)
	GOROOT/src/testing/testing.go:1690 +0xe4
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1743 +0x314
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCastMap.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/sem/eval.TestCastMap [failed](https://mesolite.cluster.engflow.com/invocations/default/910b69e0-d084-4495-8c1e-642f60e362d0?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NlbS9ldmFsOmV2YWxfdGVzdA==) on master @ [a57ed482d55a8fdfa789d6e8e1a87c2de8523f8d](https://github.com/cockroachdb/cockroach/commits/a57ed482d55a8fdfa789d6e8e1a87c2de8523f8d):

Fatal error:

```
panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize

```
Stack: 

```
goroutine 15 [running]:
testing.tRunner.func1.2({0x29732e0, 0x35484f0})
	GOROOT/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x35e
panic({0x29732e0?, 0x35484f0?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:136 +0x3aa
panic({0x29732e0?, 0x35484f0?})
	GOROOT/src/runtime/panic.go:791 +0x132
strconv.genericFtoa({0x0?, 0xfe6025?, 0x4bd128?}, 0x49e401?, 0x77?, 0xc000070c20?, 0x0?)
	GOROOT/src/strconv/ftoa.go:68 +0x685
strconv.AppendFloat(...)
	GOROOT/src/strconv/ftoa.go:54
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.PgwireFormatFloat({0x0?, 0x0, 0x63fd25?}, 0xc000b91f50?, {0x2d?, 0x0?, 0x70c78?, {0xc0?, 0x1f6116e?}}, 0xc000650bb0?)
	pkg/sql/sem/tree/pgwire_encode.go:215 +0xf5
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).pgwireFormat(0xc000813ec0, 0xc000b84300)
	pkg/sql/sem/tree/pgwire_encode.go:90 +0x1e5
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).Format(0xc0001e2190?, 0x1?)
	pkg/sql/sem/tree/datum.go:4512 +0xd2
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrAdjustConstants(0xc000b84300, {0x3554b20, 0xc000813ec0})
	pkg/sql/sem/tree/adjust_constants.go:88 +0x2e2
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0xc000b84300, {0x3554b20, 0xc000813ec0})
	pkg/sql/sem/tree/format.go:703 +0xab
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.AsStringWithFlags({0x3554b20, 0xc000813ec0}, 0x10002, {0xc000071540?, 0x10?, 0x10?})
	pkg/sql/sem/tree/format.go:841 +0x67
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCastWithoutPrecisionTruncation({0x357eab8, 0x53ad8c0}, 0xc000434008, {0x35ac720?, 0xc000813ec0?}, 0x505d320, 0x1)
	pkg/sql/sem/eval/cast.go:453 +0x6da5
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCast({0x357eab8?, 0x53ad8c0?}, 0x412?, {0x35ac720?, 0xc000813ec0?}, 0x505d320, 0x48?)
	pkg/sql/sem/eval/cast.go:100 +0x2c
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.PerformCast(...)
	pkg/sql/sem/eval/cast.go:65
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap.func1(0x2a68020?, 0x412, 0xc0?, 0xd8?, 0x22?)
	pkg/sql/sem/eval/cast_map_test.go:54 +0x11d
github.com/cockroachdb/cockroach/pkg/sql/sem/cast.ForEachCast(0xc000071f28)
	pkg/sql/sem/cast/cast.go:158 +0x386
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap(0xc0009ff040)
	pkg/sql/sem/eval/cast_map_test.go:39 +0x1dd
testing.tRunner(0xc0009ff040, 0x314b330)
	GOROOT/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1743 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestCastMap
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestCastMap1613092304
    test_log_scope.go:82: use -show-logs to present logs inline
    ftoa.go:68: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestCastMap1613092304
    ftoa.go:68: panic: strconv: illegal AppendFloat/FormatFloat bitSize
--- FAIL: TestCastMap (0.11s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCastMap.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/sem/eval.TestCastMap [failed](https://mesolite.cluster.engflow.com/invocations/default/616438e1-3f9b-43b1-8fd9-d11d102091da?testReportRun=6&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NlbS9ldmFsOmV2YWxfdGVzdA==) on master @ [223617d5f65835071f1f3395bc7b487e6dc029c9](https://github.com/cockroachdb/cockroach/commits/223617d5f65835071f1f3395bc7b487e6dc029c9):

Fatal error:

```
panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize [recovered]
	panic: strconv: illegal AppendFloat/FormatFloat bitSize

```
Stack: 

```
goroutine 15 [running]:
testing.tRunner.func1.2({0x2973920, 0x3548d50})
	GOROOT/src/testing/testing.go:1632 +0x230
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1635 +0x35e
panic({0x2973920?, 0x3548d50?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func2()
	pkg/util/leaktest/leaktest.go:136 +0x3aa
panic({0x2973920?, 0x3548d50?})
	GOROOT/src/runtime/panic.go:791 +0x132
strconv.genericFtoa({0x0?, 0xfe6025?, 0x4bd128?}, 0xc000070c08?, 0x6b?, 0xc000070c20?, 0x0?)
	GOROOT/src/strconv/ftoa.go:68 +0x685
strconv.AppendFloat(...)
	GOROOT/src/strconv/ftoa.go:54
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.PgwireFormatFloat({0x0?, 0x0, 0x63fd25?}, 0xc000be6060?, {0x2d?, 0x0?, 0x70c78?, {0xc0?, 0x1f617ae?}}, 0xc000696f88?)
	pkg/sql/sem/tree/pgwire_encode.go:215 +0xf5
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).pgwireFormat(0xc00081be00, 0xc00093e300)
	pkg/sql/sem/tree/pgwire_encode.go:90 +0x1e5
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*DTuple).Format(0x504c9e0?, 0x1?)
	pkg/sql/sem/tree/datum.go:4512 +0xd2
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).formatNodeOrAdjustConstants(0xc00093e300, {0x3555380, 0xc00081be00})
	pkg/sql/sem/tree/adjust_constants.go:88 +0x2e2
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FmtCtx).FormatNode(0xc00093e300, {0x3555380, 0xc00081be00})
	pkg/sql/sem/tree/format.go:703 +0xab
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.AsStringWithFlags({0x3555380, 0xc00081be00}, 0x10002, {0xc000071540?, 0x10?, 0x10?})
	pkg/sql/sem/tree/format.go:841 +0x67
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCastWithoutPrecisionTruncation({0x357f318, 0x53ae900}, 0xc000436008, {0x35acf80?, 0xc00081be00?}, 0x505e280, 0x1)
	pkg/sql/sem/eval/cast.go:453 +0x6da5
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.performCast({0x357f318?, 0x53ae900?}, 0x19?, {0x35acf80?, 0xc00081be00?}, 0x505e280, 0x48?)
	pkg/sql/sem/eval/cast.go:100 +0x2c
github.com/cockroachdb/cockroach/pkg/sql/sem/eval.PerformCast(...)
	pkg/sql/sem/eval/cast.go:65
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap.func1(0x2a68660?, 0x19, 0xc0?, 0xf8?, 0x22?)
	pkg/sql/sem/eval/cast_map_test.go:54 +0x11d
github.com/cockroachdb/cockroach/pkg/sql/sem/cast.ForEachCast(0xc000071f28)
	pkg/sql/sem/cast/cast.go:158 +0x386
github.com/cockroachdb/cockroach/pkg/sql/sem/eval_test.TestCastMap(0xc000a0b040)
	pkg/sql/sem/eval/cast_map_test.go:39 +0x1dd
testing.tRunner(0xc000a0b040, 0x314bb90)
	GOROOT/src/testing/testing.go:1690 +0xf4
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1743 +0x390
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestCastMap
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestCastMap1567350380
    test_log_scope.go:82: use -show-logs to present logs inline
    ftoa.go:68: -- test log scope end --

ERROR: a panic has occurred!
Details cannot be printed yet because we are still unwinding.
Hopefully the test harness prints the panic below, otherwise check the test logs.

test logs left over in: outputs.zip/logTestCastMap1567350380
    ftoa.go:68: panic: strconv: illegal AppendFloat/FormatFloat bitSize
--- FAIL: TestCastMap (0.11s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCastMap.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/randgen/datum.go
+++ b/pkg/sql/randgen/datum.go
@@ -239,16 +239,21 @@ func RandDatumWithNullChance(
 		}
 		return tree.NewDJsonpath(*jp.AST)
 	case types.TupleFamily:
+		tuple := tree.DTuple{D: make(tree.Datums, len(typ.TupleContents()))}
 		if nullChance == 0 {
+			// Even if nullOk=false (which is when nullChance=0), we still want
+			// to generate a NULL _element_ in 10% of cases.
 			nullChance = 10
 		}
-		datums := make([]tree.Datum, len(typ.TupleContents()))
 		for i := range typ.TupleContents() {
-			datums[i] = RandDatumWithNullChance(
+			tuple.D[i] = RandDatumWithNullChance(
 				rng, typ.TupleContents()[i], nullChance, favorCommonData, targetColumnIsUnique,
 			)
 		}
-		return tree.NewDTuple(typ, datums...)
+		// Calling ResolvedType causes the internal TupleContents types to be
+		// populated (as well as resolves the type of each tuple element).
+		tuple.ResolvedType()
+		return &tuple
 	case types.BitFamily:
 		width := typ.Width()
 		if width == 0 {

--- a/pkg/sql/randgen/types_test.go
+++ b/pkg/sql/randgen/types_test.go
@@ -9,6 +9,7 @@ import (
 	"fmt"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
@@ -66,7 +67,30 @@ func TestCanonical(t *testing.T) {
 		datum := RandDatum(rng, typ, false)
 		datumTyp := datum.ResolvedType()
 		if !datumTyp.Equivalent(typ.Canonical()) {
-			t.Errorf("fail: canonical type of %+v is %+v and the datum's type is %+v", typ, typ.Canonical(), datumTyp)
+			var tupleHasNullElement func(tree.Datum) bool
+			tupleHasNullElement = func(d tree.Datum) bool {
+				tup, ok := d.(*tree.DTuple)
+				if !ok {
+					return false
+				}
+				for _, el := range tup.D {
+					if el == tree.DNull {
+						return true
+					}
+					if tupleHasNullElement(el) {
+						return true
+					}
+				}
+				return false
+			}
+			// If we have a tuple, then we might have included a NULL element.
+			// By construction in RandDatum, TupleContents[i] has been updated
+			// to types.Unknown. In such case, we give an exception and don't
+			// fail the test.
+			tupleException := tupleHasNullElement(datum)
+			if !tupleException {
+				t.Errorf("fail: canonical type of %+v is %+v and the datum's type is %+v", typ, typ.Canonical(), datumTyp)
+			}
 		}
 
 		if datumTyp.Oid() != typ.Canonical().Oid() {

